### PR TITLE
vim-patch:8.1.{1418,1425,1832,2124},8.2.{0137, 2340}

### PR DESCRIPTION
**Pull request by ivechan**
To port Vim's win_execute() function.  https://github.com/neovim/neovim/issues/10822
extra tests should be added into tests/functinal/ui/float_spec.lua

### CODE REVIEW

**Comment by bfredl** [comment]
nice! We should also add `nvim_win_call` for lua in analogy with `nvim_buf_call`, but I will do that as a follow-up PR.

File: src/nvim/window.c [hunk 1]
```diff
@@ -6528,6 +6533,16 @@ static win_T *get_snapshot_focus(int idx)
 int switch_win(win_T **save_curwin, tabpage_T **save_curtab, win_T *win, tabpage_T *tp, int no_display)
 {
   block_autocmds();
+  return switch_win_noblock(save_curwin, save_curtab, win, tp, no_display);
+}
+
+// As switch_win() but without blocking autocommands.
+int switch_win_noblock(win_T **save_curwin,
```
Comment (bfredl) [inline src/nvim/window.c:6540]: we should also use the "noblock" variants in `nvim_win_set_buf`

**Comment by bfredl** [comment]
I think this should be in 0.5 as it will help fix some bugs :]

**Comment by ivechan** [comment]
@bfredl I think current legacy tests in this PR is enough.
But I can't figure out why CI / windows (MSVC_32) failed because the error on MSVC_32 seems to be not related to this PR.

**Comment by ivechan** [comment]
This PR is ready to be merged if you agreed.

File: src/nvim/window.c (hunk 1 above)
Comment (ivechan) [reply src/nvim/window.c:6540]: done! and add a testcase in test/functional/autocmd/window_spec.lua .

**Comment by ivechan** [comment]
now autocmd can work as expected when calling nvim_win_set_buf.
I added a testcase to make sure it work fine.

**Comment by janlazo** [comment]
Commit format on some commits and the PR title are wrong. It's causing the PR to not be auto-labeled with the `vim-patch` label.

**Comment by ivechan** [comment]
> Commit format on some commits and the PR title are wrong. It's causing the PR to not be auto-labeled with the `vim-patch` label.

I apologize it.
I have corrected the format and resubmitted, please review it agian :)

File: runtime/doc/eval.txt [hunk 2]
```diff
@@ -9262,6 +9262,12 @@ wildmenumode()					*wildmenumode()*
 <
 		(Note, this needs the 'wildcharm' option set appropriately).
 
+win_execute({id}, {command} [, {silent}])		*win_execute()*
```
Comment (janlazo) [inline runtime/doc/eval.txt:9265]: This is not the only change in this file.

File: src/nvim/eval/funcs.c [hunk 3]
```diff
@@ -1982,6 +1986,37 @@ static void f_execute(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   capture_ga = save_capture_ga;
 }
 
+/// "execute()" function
+static void f_execute(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+{
+    execute_common(argvars, rettv, fptr, 0);
```
Comment (janlazo) [inline src/nvim/eval/funcs.c:1992]: 2-space indent.

File: src/nvim/testdir/test_execute_func.vim [hunk 4]
```diff
@@ -82,3 +84,43 @@ func Test_execute_not_silent()
...
+
+func Test_win_execute()
+  let thiswin = win_getid()
+  new
+  let otherwin = win_getid()
+  call setline(1, 'the new window')
+  call win_gotoid(thiswin)
+  let line = win_execute(otherwin, 'echo getline(1)')
+  call assert_match('the new window', line)
+
```
Comment (janlazo) [inline src/nvim/testdir/test_execute_func.vim:96]: https://github.com/vim/vim/commit/868b7b6712ea4f2232eeeae18c5cbbbddf2ee84d#diff-2fa6230b9e8601cd9e1fae2257f1c95f95e307ede63335c924646f90582eee34R91-R100 is missing.

Do not remove Vimscript code in new-style Vim tests that should be skipped in Neovim.

**Comment by ivechan** [comment]
WIP, many issue need to be address.

**Comment by ivechan** [comment]
I re-finished this PR
And this time I was  careful to avoid unexpected inconsistencies with the original vim patches.

text_prop, popup and vim9 related changes are not introduced to src/

**Comment by janlazo** [comment]
Adding back to 0.5.0 if it's ready because it's mentioned in  https://github.com/neovim/neovim/pull/14424. I shouldn't have done that but I don't want to amend any commits there.

File: src/nvim/window.c [hunk 5]
```diff
@@ -6359,6 +6370,13 @@ int switch_win(win_T **save_curwin, tabpage_T **save_curtab, win_T *win, tabpage
 // When "no_display" is true the display won't be affected, no redraw is
 // triggered.
 void restore_win(win_T *save_curwin, tabpage_T *save_curtab, bool no_display)
+{
+  restore_win_noblock(save_curwin, save_curtab, no_display);
+  unblock_autocmds();
+}
+
+void restore_win_noblock(win_T *save_curwin, tabpage_T *save_curtab,
```
Comment (janlazo) [inline src/nvim/window.c:6378]: Function description missing.

File: src/nvim/window.c [hunk 6]
```diff
@@ -6335,6 +6340,12 @@ static win_T *get_snapshot_focus(int idx)
 int switch_win(win_T **save_curwin, tabpage_T **save_curtab, win_T *win, tabpage_T *tp, int no_display)
 {
   block_autocmds();
+  return switch_win_noblock(save_curwin, save_curtab, win, tp, no_display);
+}
+
+int switch_win_noblock(win_T **save_curwin, tabpage_T **save_curtab,
```
Comment (janlazo) [inline src/nvim/window.c:6346]: Function description missing.

**Comment by janlazo** [comment]
No lua tests? Vim's popup code is not ported so some test cases here don't run in the CI.

**Review by janlazo — CHANGES_REQUESTED** [review]
LGTM aside from my comments.

**Comment by ivechan** [comment]
> No lua tests? Vim's popup code is not ported so some test cases here don't run in the CI.
What test cases do you think need to be added?
I plan to port the testcases of .vim to .lua and replace popup related testcases with float window testcases。

**Comment by janlazo** [comment]
> What test cases do you think need to be added?

Any tests you already have for `tests/functinal/ui/float_spec.lua`.
Optionally, any Vim screendump tests that you ported to a Neovim screen test.

**Comment by ivechan** [comment]
The function descriptions were added.
And I just added two basic testcases in `test/functional/ui/float_spec.lua`

**Review by janlazo — APPROVED** [review]
LGTM

**Comment by clason** [comment]
Just to note this here: this PR broke telescope; trying to open a finder throws
```
E5108: Error executing lua .../packer/start/plenary.nvim/lua/plenary/window/border.lua:138: Vim(autocmd):E680: <buffer=6>: invalid buffer number
```
If anybody here has a quick idea what needs to be changed in telescope to make it work with the changed code, that would be helpful.

**Comment by kevinhwang91** [comment]
After the patch, `nvim_open_win` call `nvim_win_set_buf` and throw error. Please give a correct way for using `nvim_open_win`.

**Comment by ray-x** [comment]
I am not 100% sure what this means to autocommand
The fix for telescope is removing "BufLeave" from autocommand.  
But what happens if the plugin relay on "BufLeave"? Is there a new API to set autocommand?

**Comment by tomtomjhj** [comment]
I don't quite understand https://github.com/neovim/neovim/pull/13664/commits/1def3d1542d6a65f057e743faea39a760b50db87. This commit makes `nvim_win_set_buf` not block autocmds so that it's more "without side-effects", but I think allowing autocmds results in more side-effects. Has anyone tried reverting this commit? My setup is not affected by this PR, so I can't test myself.

**Comment by ddickstein** [comment]
@tomtomjhj I may be missing some context here but I think that when a user expects to run certain logic every time Foo happens (expressed by registering an autocmd) and then you trigger Foo but prevent that logic from running (noautocmd), you are circumventing the user's intention / doing something surprising. Maybe that's what the author means by "without side-effects" -- without special behavior. But in the literal sense it's definitely more "effectful" to run logic on events than to not run logic on events.

**Comment by janlazo** [comment]
Created https://github.com/neovim/neovim/pull/14531 to revert the breaking changes if there's no fix within this month.

**Comment by bfredl** [comment]
I would rather add a `noautocmd=true` flag to `nvim_open_win()`. Running Enter autocmds for the specific buffer-window pair is a __main__ effect of showing that buffer in the window, and needed to support filetypes, cmd buffers and a lot of other stuff (the side effects we avoid is unnecessary winenter/leave autocmds of _other_ windows which a naive user-space re-implementation of `nvim_win_set_buf` would could cause)

**Comment by ivechan** [comment]
What @ddickstein said explains what I meant by  "without side-effects".


> Is there a missing +popupwin patch that should be ported to Neovim's floating window?

I will check it.

> I would rather add a noautocmd=true flag to nvim_open_win().

Are there any scenarios where we would expect nvim_win_set_buf to trigger autocmd? 
If not, I think it might be better to just revert this commit

**Comment by bfredl** [comment]
> Are there any scenarios where we would expect nvim_win_set_buf to trigger autocmd?
If not, I think it might be better to just revert this commit

The same scenarios where `:edit` and `:buffer` is expected to trigger autocmds, of which I listed some above. Most importantly filetype (and bufreadcmd), which is now only supported in `nvim_open_win` through hacks.  Editing a buffer in a floating window should generally be expected to behave the same as using another window, and it is much simpler and reliable to disable autocmds when we don't want them that faking them where they got missing.

**Pull request merged**